portal-kernel/src/com/liferay/portal/kernel/instance/PortalInstancePool.java

```diff
@@ -23,35 +33,214 @@ public static void add(Company company) {
 		_portalInstances.put(company.getCompanyId(), company.getWebId());
 	}
 
+	public static void add(List<Company> companies) {
+		for (Company company : companies) {
+			_portalInstances.put(company.getCompanyId(), company.getWebId());
+		}
+	}
+
+	public static long getCompanyId(String webId) {
+		if (!_portalInstances.isEmpty()) {
+			for (Map.Entry<Long, String> entry : _portalInstances.entrySet()) {
+				if (Objects.equals(entry.getValue(), webId)) {
+					return entry.getKey();
+				}
+			}
+
+			throw new IllegalArgumentException(
+				"Unable to get company ID with web ID" + webId);
+		}
+
+		try {
+			return _getCompanyIdBySQL(webId);
+		}
+		catch (SQLException sqlException) {
+			_log.error(
+				"Unable to get the company ID for webId " + webId + " by SQL",
+				sqlException);
+
+			throw new RuntimeException(sqlException);
+		}
+	}
+
 	public static long[] getCompanyIds() {
-		return ArrayUtil.toLongArray(_portalInstances.keySet());
+		if (!_portalInstances.isEmpty()) {
```

Comment:
Hi @shuyangzhou,

That's why we changed the way we load the companies. Now we load all the companies at once:
https://github.com/shuyangzhou/liferay-portal/pull/11347/commits/9f14f6b5ad7b880195f184112b7a91c8c51396f2#diff-56eeb87f080f1f19b598b575d69314e812867eac7dc7b872d4f272f150a72633R811

Reply:
No, that's not enough. It only makes the race window smaller, not eliminating it. The `PortalInstancePool.add(companies)` is not an atomic operation either. To me it is actually making things worse. Because if it was only added one by one, when race happens it will easy to identify and reproduce. But now, it becomes a rare race that is going to very hard to identify and reproduce.

Reply:
@shuyangzhou, we made the operation atomic so we won't have this issue:
https://github.com/shuyangzhou/liferay-portal/blob/f5430457cbde67b19078795b69f781822c9f4137/portal-kernel/src/com/liferay/portal/kernel/instance/PortalInstancePool.java#L138